control-plane/api-gateway/controllers/gateway_controller.go

```diff
@@ -0,0 +1,260 @@
+package controllers
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gwv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	gatewayFinalizer = "gateway-finalizer.consul.hashicorp.com"
+
+	kindGateway = "Gateway"
+)
+
+// GatewayController reconciles a Gateway object.
+// The Gateway is responsible for defining the behavior of API gateways.
+type GatewayController struct {
+	Log logr.Logger
+	client.Client
+}
+
+// Reconcile handles the reconciliation loop for Gateway objects.
+func (r *GatewayController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gateway", req.NamespacedName)
+	log.Info("Reconciling the Gateway: ", req.Name)
+
+	// If gateway does not exist, log an error.
+	gw := &gwv1beta1.Gateway{}
+	err := r.Client.Get(ctx, req.NamespacedName, gw)
+	if err != nil {
+		log.Error(err, "unable to get Gateway")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If gateway class on the gateway does not exist, log an error.
+	gwc := &gwv1beta1.GatewayClass{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: string(gw.Spec.GatewayClassName)}, gwc)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, err
+	}
+
+	if string(gwc.Spec.ControllerName) != GatewayClassControllerName || !gw.ObjectMeta.DeletionTimestamp.IsZero() {
+		// This Gateway is not for this controller or the gateway is being deleted.
+		// TODO: Cleanup Consul resources.
```

Comment:
This is where we'll stick the `gatekeeper.DeleteGateway` call and also something else to inform Consul.